### PR TITLE
Simplify the DualPipeV pipeline engine

### PR DESCRIPTION
**Pull request by haok1402**
# Simplify the DualPipeV pipeline engine

## Why

`pithtrain/dualpipe/` was named for the upstream project it vendors, sitting at the same level as role-named packages like `operators/` and `contexts/`. The dependency graph says it is not a peer of `modules/` at all: all four model files and two operator files import it, while `modules/training.py` consumes it. It is a low layer wearing a vendored name.

Three concrete defects came out of reading it:

- The `add-new-model` skill instructed agents to `from ...execution import record_forward`. No such symbol exists — it is `model_forward`. Any agent following the skill wrote a model file that fails at import, including the copy-pasted `templates/model_skeleton.py`.
- `NOTICE` attributed derived code to `pithtrain/dualpipe/comm.py`, a file merged into `dualpipev.py` two PRs ago, and named two functions deleted since. That is a legal attribution file pointing at a path that does not exist.
- Most of `utils.py` had no callers anywhere in the repo.

The timing matters: every consumer of these paths is still in this repository, so the rename is a local change. Once an external training backend imports them, it stops being one.

## Changes

Five commits, each independently testable. The package goes from four source files to two.

| commit | change |
|---|---|
| `2d31d96` | Delete the dead utilities; fix `NOTICE`; fix `record_forward` -> `model_forward` across the skill and its template |
| `4518f52` | Dissolve `utils.py` — derived code into `execution.py`, `FP8WeightCacheControl` down into `operators/` |
| `f52b661` | Merge `overlap.py` into `dualpipev.py` — one caller, and the file boundary forced a 13-parameter signature |
| `47aab54` | Rename `pithtrain/dualpipe/` -> `pithtrain/pipeline/` |
| `73598f4` | Drop dead attributes and parameters, stop threading `ep_group`, collapse the merged-layer predicate, unify gradient naming |

No arithmetic changed. The one rename that could have crossed wires, `stage1_stage5_b`'s `prev_layer`/`next_layer`, is read only by the NVTX label.

## Correctness

`bash tests/test_dualpipev.sh` builds each model at `phase=-1` (single-device reference) and `phase=0/1` (the two DualPipeV chunks), then compares the pipelined 5-stage forward against the reference on loss and on every parameter gradient. This is a tighter check than a loss curve.

| model | dense | packed (`PACKED_SEQLEN=1`) | max gradient diff |
|---|---|---|---|
| deepseek-v2-lite | PASS | PASS | 7.1e-4 |
| qwen3-30b-a3b | PASS | PASS | 5.1e-4 |
| gpt-oss-20b | PASS | PASS | 1.3e-3 |
| qwen3.5-35b-a3b | PASS | n/a (asserts dense) | — |

All against a `calc_diff` bound of 1e-2. Operator tests: 22 passed, 2 failed — both failures reproduced identically at the parent commit, so they are pre-existing.

**FP8.** `test_dualpipev.py` hardcodes `training.fp8 = False`, so the FP8 path is untested upstream — and this branch moves `FP8WeightCacheControl`, whose only callers are `DualPipeV.step()`. Verified manually with a temporary patch binding the FP8 backends: **gpt-oss-20b passes** on loss and gradients. qwen3-30b-a3b cannot run FP8 in this harness for an unrelated pre-existing reason: `shard_experts` detects experts via `isinstance(child, GroupedLinear)`, which `FP8GroupedLinear` fails, so experts are never sharded (128 vs 64). Worth a follow-up; not a regression here.

**Loss curve.** `validate-correctness` on qwen3-30b-a3b at pp=2 ep=8 cp=1, 4k sequence, 512 global batch, 32 steps, real routing (`benchmark=False`).

```
cross-entropy-loss     envelope 0.0226  signal 0.0504  ratio  2.23  PASS
load-balance-loss      envelope 0.0529  signal 0.1005  ratio  1.90  PASS
step-1 forward         envelope 0.0000  signal 0.0001
```

| run | branch | wandb |
|---|---|---|
| base0 | `47aab54` | [oplqgbpe](https://wandb.ai/PithTrain/0904-dualpipe-cleanup/runs/oplqgbpe) |
| base1 | `47aab54` | [8jop6mwq](https://wandb.ai/PithTrain/0904-dualpipe-cleanup/runs/8jop6mwq) |
| feat0 | `73598f4` | [245ivl1i](https://wandb.ai/PithTrain/0904-dualpipe-cleanup/runs/245ivl1i) |

The yardstick is base-vs-base, not zero, because two runs of *identical code* do not agree: final loss 9.5154 vs 9.5140, and final gradient norm 2.6492 vs 5.3850. FlashAttention-4's backward is not reproducible even under `use_deterministic_algorithms(True)`. Both gated metrics land at roughly 2x that floor against a 3x pass line.

## Performance

Separate pass with `training.benchmark = True`, same mesh, comparing run medians rather than pooled steps.

| run | branch | median step-time | wandb |
|---|---|---|---|
| perf-base0 | `47aab54` | 11.584 s | [3pjbiybk](https://wandb.ai/PithTrain/0904-dualpipe-cleanup/runs/3pjbiybk) |
| perf-base1 | `47aab54` | 11.611 s | [v04jsypy](https://wandb.ai/PithTrain/0904-dualpipe-cleanup/runs/v04jsypy) |
| perf-feat0 | `73598f4` | 11.652 s | [mz34u4pc](https://wandb.ai/PithTrain/0904-dualpipe-cleanup/runs/mz34u4pc) |
| perf-feat1 | `73598f4` | 11.607 s | [okpm7mdx](https://wandb.ai/PithTrain/0904-dualpipe-cleanup/runs/okpm7mdx) |

Feature median is +0.28% against a base-vs-base floor of 0.23%, and `feat1` (11.607) falls *inside* the base range (11.584-11.611). The arms interleave, so there is no separation to claim. Over the first 8 steps the floor is 0.68% against a delta of 0.70% — likewise indistinguishable. Two reps per arm, which the forced-balanced routing makes sufficient.

### CODE REVIEW

**Comment by haok1402** [comment]
@claude review
